arch/riscv/core/userspace.S

```diff
@@ -0,0 +1,78 @@
+/*
+ * Userspace and service handler hooks
+ *
+ * Copyright (c) 2020 BayLibre, SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <offsets_short.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+#include <kernel_structs.h>
+#include <arch/riscv/csr.h>
+
+/* exports */
+GTEXT(z_riscv_do_syscall)
+GTEXT(arch_user_string_nlen)
+GTEXT(z_riscv_user_string_nlen_fault_start)
+GTEXT(z_riscv_user_string_nlen_fault_end)
+GTEXT(z_riscv_user_string_nlen_fixup)
+
+/* Imports */
+GDATA(_k_syscall_table)
+
+SECTION_FUNC(TEXT,z_riscv_do_syscall)
+	la	t0, _k_syscall_table
+
+	slli	t1, a7, RV_REGSHIFT	# Determine offset from indice value
+	add	t0, t0, t1		# Table addr + offset = function addr
+	RV_OP_LOADREG t3, 0x00(t0)	# Load function address
+
+	/* Allocate space on thread stack to save registers */
+	addi sp,sp, -RV_REGSIZE
+
+	/* Save RA registers on current thread stack. */
+	RV_OP_STOREREG ra, __z_arch_esf_t_ra_OFFSET(sp)
+
+	/* Execute syscall function */
+	jalr 	t3
+
+	/* Restore RA registers from thread stack */
+	RV_OP_LOADREG ra, __z_arch_esf_t_ra_OFFSET(sp)
```

Comment:
At the beginning, I over-wrote all regs before and after syscall jump. But after test/reflexion, I think it's overkill to do that and a loss of performance. Back from syscall, I only save a0 in the existing ESF. The other regs will not be directly used with there current value. Do you think I should "flush" regs before jump to the syscall function ?

Reply:
> Do you think I should "flush" regs before jump to the syscall function

when you come back from the system call back to the calling user thread, there's nothing in any registers that was put there when the system call was being handled by the kernel. 

So in the chain of events:

1. Invoke system call
2. Marshal arguments and initiate SW trap
3. Handle SW exception, look up system call ID
4. Invoke system call inside kernel from dispatch table
5. Marshal return value and exit SW trap back to caller

No register values from steps 3-5 should leak back to user mode with the exception of the return value.

Reply:
Then it's ok because during step 5, I save return value, jump back in isr.S to finish syscall handler, that mean all register values are restored (exept return register) from saved value in step 1

Reply:
perfect